test/ecto/adapters/postgres_test.exs

```diff
@@ -883,6 +883,51 @@ defmodule Ecto.Adapters.PostgresTest do
     """ |> remove_newlines]
   end
 
+  test "create table with binary column and null-byte default" do
+    create = {:create, table(:blobs),
+              [{:add, :blob, :binary, [default: <<0>>]}]}
+
+    assert_raise ArgumentError, ~r"UTF-8", fn ->
+      execute_ddl(create)
+    end
+  end
+
+  test "create table with binary column and null-byte-containing default" do
+    create = {:create, table(:blobs),
+              [{:add, :blob, :binary, [default: "foo" <> <<0>>]}]}
+
+    assert_raise ArgumentError, ~r"UTF-8", fn ->
```

Comment:
If we do my suggestion above, let's check the suggestion in the error message is correct. :)